Userland/DynamicLoader/main.cpp

```diff
@@ -36,18 +38,75 @@ static void perform_self_relocations(auxv_t* auxvp)
         exit(1);
 }
 
-static void display_help()
+static ErrorOr<int> open_executable(StringView path)
 {
-    char const message[] =
-        R"(You have invoked `Loader.so'. This is the helper program for programs that
-use shared libraries. Special directives embedded in executables tell the
-kernel to load this program.
-
-This helper program loads the shared libraries needed by the program,
-prepares the program to run, and runs it. You do not need to invoke
-this helper program directly. If you still want to run a program with the loader,
-run: /usr/lib/Loader.so [ELF_BINARY])";
-    fprintf(stderr, "%s", message);
+    int rc = open(path.characters_without_null_termination(), O_RDONLY | O_EXEC);
+    if (rc < 0)
+        return Error::from_errno(errno);
+    int checked_fd = rc;
+    ArmedScopeGuard close_on_failure([checked_fd] {
+        close(checked_fd);
+    });
+
+    struct stat executable_stat = {};
+    rc = fstat(checked_fd, &executable_stat);
+    if (rc < 0)
+        return Error::from_errno(errno);
+    if (!S_ISREG(executable_stat.st_mode)) {
+        if (S_ISDIR(executable_stat.st_mode))
+            return Error::from_errno(EISDIR);
+        return Error::from_errno(EINVAL);
+    }
+
+    close_on_failure.disarm();
+    return checked_fd;
+}
+
+static int _main(int argc, char** argv, char** envp, bool is_secure)
+{
+    tzset();
+
+    Vector<StringView> arguments;
+    arguments.ensure_capacity(argc);
+    for (int i = 0; i < argc; ++i)
+        arguments.unchecked_append({ argv[i], strlen(argv[i]) });
+
+    Vector<StringView> command;
+    Core::ArgsParser args_parser;
+
+    Function<void(int)> exit_callback(&_exit);
```

Comment:
Can we instead make exit work?

Reply:
I could try to debug why LibC `exit` does not work. It has a nullptr deref somehow, probably because of how it tries to finalize things that were never initialized in the first place? why is that a problem to us anyway?

Reply:
So I changed things so we will fail with print usage and then I check the boolean result of `ArgsParser::parse` method, and if it fails I simply exit on my own terms. Seems like a legitimate option here, which doesn't modify `ArgsParser` code anymore.